src/ripple/overlay/impl/ProtocolVersion.cpp

```diff
@@ -0,0 +1,140 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2019 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/overlay/impl/ProtocolVersion.h>
+#include <ripple/beast/core/LexicalCast.h>
+#include <ripple/beast/rfc2616.h>
+#include <boost/regex.hpp>
+#include <algorithm>
+
+namespace ripple {
+
+/** The list of protocol versions we speak and we prefer to use. */
+constexpr
+ProtocolVersion const supportedProtocolList[]
+{
+    { 1, 2 },
+    { 2, 0 }
+};
+
+constexpr
+ProtocolVersion const legacyProtocol { 1, 2 };
+
+std::string
+to_string(ProtocolVersion const& p)
+{
+    if (p == legacyProtocol)
+        return "RTXP/" + std::to_string(p.first) + "." + std::to_string(p.second);
+
+    return "XRPL/" + std::to_string(p.first) + "." + std::to_string(p.second);
+}
+
+std::vector<ProtocolVersion>
+parseProtocolVersions(boost::beast::string_view const& value)
+{
+    static boost::regex re(
+        "^"                  // start of line
+        "(?:"                // Alternative #1: old-style
+        "RTXP/(1)\\.(2)"     // The string "RTXP/1.2"
+        "|"                  // Alternative #2: new-style
+        "XRPL/"              // the string "XRPL/"
+        "([2-9][0-9]*)"      // a number (greater than 2 with no leading zeroes)
+        "\\."                // a period
+        "(0|[1-9][0-9]*)"    // a number (no leading zeroes unless exactly zero)
+        ")"
+        "$"                  // The end of the string
+        , boost::regex_constants::optimize);
+
+    static std::string const legacy = to_string(legacyProtocol);
+
+    std::vector<ProtocolVersion> result;
+
+    for (auto const& s : beast::rfc2616::split_commas(value))
```

Comment:
There's no error handling here. For example, if we pass a list of versions with "junk on the end" it ignores it. Is this OK?

Reply:
I'm not sure I follow. This code will crack a given input string at the commas, and for each item, it will call `boost::regex_match` and we only do work if the item matches the regex description. Only two kinds of entries will pass:

1. **`RTXP/1.2`**
2. **`XRPL/x.y`** where `x ≥ 2` and `y ≥ 0`.

Reply:
I was considering the case: `RTXP/1.2, GOBBLYGOOK`. But now that I think about it, of course you want to silently ignore `GOBBLYGOOK`.

Reply:
What about `XRPL/10.0`?

Reply:
@MarkusTeufelberger I agree, that looks broken. Good find!

Reply:
I'm not sure I follow... **`XRPL/10.0`** is theoretically a valid protocol to request. The regular expression accounts for it:

    "XRPL/"              // the string "XRPL/"
    "([2-9][0-9]*)"      // a number (greater than 2 with no leading zeroes)
    "\\."                // a period
    "(0|[1-9][0-9]*)"    // a number (no leading zeroes unless exactly zero)

So, a string like `RTXP/1.2, GOBBLYGOOK, XRPL/2.0, XRPL/10.0` will, correctly, produce the following list: **`{ { 1, 2 }, { 2, 0 }, { 10, 0 } }`**

Reply:
Fixed by upgrading the regex from:
`([2-9][0-9]*)` to:
`([2-9]|[1-9][0-9]+)`

Thanks @MarkusTeufelberger.